File: web-app/src/types/telemetry.ts
export type Vec2 = [number, number];

export interface EntityState {
  id: number;
  position: Vec2;
  velocity: Vec2;
  health: number;
  max_health: number;
  team: number;
  is_dead: boolean;
  facing: number;
}

export type TelemetryEvent =
  | { type: 'WorldSnapshot'; tick: number; entities: EntityState[] }
  | { type: 'Kill'; tick: number; killer: number; victim: number }
  | { type: 'Damage'; tick: number; source: number; target: number; amount: number }
  | { type: 'ShotFired'; tick: number; shooter: number; origin: Vec2; direction: Vec2; hit_target: number | null }
  | { type: 'Spawn'; tick: number; entity: number; position: Vec2; team: number }
  | { type: 'TickComplete'; tick: number; entity_count: number }
  | { type: 'RoundStart'; tick: number; obstacles: Array<{ x: number; y: number; width: number; height: number }>; spawn_points: Vec2[] }
  | { type: 'TacticalState'; tick: number; entity: number; move_target: number; candidates: [number, number][]; candidate_los: boolean[]; path: [number, number][]; aim_angle: number; shooting: boolean; ray_distances: number[]; rewards?: RewardBreakdown }
  | { type: 'Arena3DState'; tick: number; entity: number; position: [number, number, number]; velocity: [number, number, number]; yaw: number; pitch: number; health: number; max_health: number; team: number; is_dead: boolean; active_weapon: number; shooting: boolean; move_direction: number; ray_distances: number[]; ray_hit_types: number[] }
  | { type: 'CsLiteRoundState'; tick: number; phase: string; round_number: number; t_score: number; ct_score: number; phase_timer: number; t_alive: number; ct_alive: number };

export interface RewardBreakdown {
  proximity: number;
  aim: number;
  time_penalty: number;
  cover_bonus: number;
  idle_penalty: number;
  los_gain: number;
  combat: number;
}

File: crates/engine/src/telemetry.rs
use glam::Vec2;
use serde::{Deserialize, Serialize};

use crate::ecs::components::ObstacleRect;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EntityState {
    pub id: u64,
    pub position: Vec2,
    pub velocity: Vec2,
    pub health: f32,
    pub max_health: f32,
    pub team: u8,
    pub is_dead: bool,
    pub facing: f32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type")]
pub enum TelemetryEvent {
    WorldSnapshot {
        tick: u64,
        entities: Vec<EntityState>,
    },
    Damage {
        tick: u64,
        source: u64,
        target: u64,
        amount: f32,
    },
    Kill {
        tick: u64,
        killer: u64,
        victim: u64,
    },
    Spawn {
        tick: u64,
        entity: u64,
        position: Vec2,
        team: u8,
    },
    ShotFired {
        tick: u64,
        shooter: u64,
        origin: Vec2,
        direction: Vec2,
        hit_target: Option<u64>,
    },
    RoundStart {
        tick: u64,
        obstacles: Vec<ObstacleRect>,
        spawn_points: Vec<Vec2>,
    },
    TickComplete {
        tick: u64,
        entity_count: usize,
    },
    TacticalState {
        tick: u64,
        entity: u64,
        move_target: u8,
        candidates: Vec<[f32; 2]>,
        candidate_los: Vec<bool>,
        path: Vec<[f32; 2]>,
        aim_angle: f32,
        shooting: bool,
        ray_distances: Vec<f32>,
        #[serde(skip_serializing_if = "Option::is_none")]
        rewards: Option<RewardBreakdown>,
    },
    Arena3DState {
        tick: u64,
        entity: u64,
        position: [f32; 3],
        velocity: [f32; 3],
        yaw: f32,
        pitch: f32,
        health: f32,
        max_health: f32,
        team: u8,
        is_dead: bool,
        active_weapon: u8,
        shooting: bool,
        move_direction: u8,
        ray_distances: Vec<f32>,
        ray_hit_types: Vec<f32>,
    },
    CsLiteRoundState {
        tick: u64,
        phase: String,
        round_number: u32,
        t_score: u32,
        ct_score: u32,
        phase_timer: f32,
        t_alive: u32,
        ct_alive: u32,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RewardBreakdown {
    pub proximity: f32,
    pub aim: f32,
    pub time_penalty: f32,
    pub cover_bonus: f32,
    pub idle_penalty: f32,
    pub los_gain: f32,
    pub combat: f32,
}
